src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -459,8 +481,459 @@ public void WithAnalyzerConfigDocumentTextLoader()
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectOutputRefFilePath(ProjectId.CreateNewId(), "x.dll"));
+        }
+
+        [Fact]
+        public void WithProjectDefaultNamespace()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            // any character is allowed
+            var defaultNamespace = "\0<>a/b/*";
+
+            var newSolution = solution.WithProjectDefaultNamespace(projectId, defaultNamespace);
+            Assert.Equal(defaultNamespace, newSolution.GetProject(projectId)!.DefaultNamespace);
+
+            Assert.Same(newSolution, newSolution.WithProjectDefaultNamespace(projectId, defaultNamespace));
+
+            var newSolution2 = solution.WithProjectDefaultNamespace(projectId, null);
+            Assert.Null(newSolution2.GetProject(projectId)!.DefaultNamespace);
+
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectDefaultNamespace(null!, "x"));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectDefaultNamespace(ProjectId.CreateNewId(), "x"));
+        }
+
+        [Fact]
+        public void WithProjectName()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            // any character is allowed
+            var projectName = "\0<>a/b/*";
+
+            var newSolution = solution.WithProjectName(projectId, projectName);
+            Assert.Equal(projectName, newSolution.GetProject(projectId)!.Name);
+
+            Assert.Same(newSolution, newSolution.WithProjectName(projectId, projectName));
+
+            Assert.Throws<ArgumentNullException>("name", () => solution.WithProjectName(projectId, null!));
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectName(null!, "x"));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectName(ProjectId.CreateNewId(), "x"));
+        }
+
+        [Fact]
+        public void WithProjectFilePath()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            // any character is allowed
+            var path = "\0<>a/b/*.csproj";
+
+            var newSolution = solution.WithProjectFilePath(projectId, path);
+            Assert.Equal(path, newSolution.GetProject(projectId)!.FilePath);
+
+            Assert.Same(newSolution, newSolution.WithProjectFilePath(projectId, path));
+
+            var newSolution2 = solution.WithProjectFilePath(projectId, null);
+            Assert.Null(newSolution2.GetProject(projectId)!.DefaultNamespace);
```

Comment:
I have caught you copy/pasting. :smile:

Reply:
(This is a really good argument for a generic "test the pass-through string" test helper. Take an optional boolean if nulls are allowed or not.)